src/cascadia/WindowsTerminal/AppHost.cpp

```diff
@@ -939,3 +957,222 @@ void AppHost::_SummonWindowRequested(const winrt::Windows::Foundation::IInspecta
+        MENUINFO mi{};
+        mi.cbSize = sizeof(MENUINFO);
+        mi.fMask = MIM_STYLE | MIM_APPLYTOSUBMENUS | MIM_MENUDATA;
+        mi.dwStyle = MNS_NOTIFYBYPOS;
+        mi.dwMenuData = NULL;
+        SetMenuInfo(hmenu, &mi);
+
+        // Focus Current Terminal Window
+        AppendMenu(hmenu, MF_STRING, (UINT_PTR)TrayMenuItemAction::FocusTerminal, RS_(L"TrayIconFocusTerminal").c_str());
+        AppendMenu(hmenu, MF_SEPARATOR, 0, L"");
+
+        // Submenu for Windows
+        if (auto windowSubmenu = _CreateWindowSubmenu())
+        {
+            MENUINFO submenuInfo{};
+            submenuInfo.cbSize = sizeof(MENUINFO);
+            submenuInfo.fMask = MIM_MENUDATA;
+            submenuInfo.dwStyle = MNS_NOTIFYBYPOS;
+            submenuInfo.dwMenuData = (UINT_PTR)TrayMenuItemAction::SummonWindow;
+            SetMenuInfo(windowSubmenu, &submenuInfo);
+
+            AppendMenu(hmenu, MF_POPUP, (UINT_PTR)windowSubmenu, RS_(L"TrayIconWindowSubmenu").c_str());
+        }
+    }
+    return hmenu;
+}
+
+// Method Description:
+// - Create a menu with a menu item for each window available to summon.
+//   If a window is unnamed, we'll use its ID but still mention that it's unnamed.
+// Arguments:
+// - <none>
+// Return Value:
+// - The handle to the newly created window submenu.
+HMENU AppHost::_CreateWindowSubmenu()
+{
+    if (auto hmenu = CreatePopupMenu())
+    {
+        auto locWindow = RS_(L"WindowIdLabel");
+        auto locUnnamed = RS_(L"UnnamedWindowName");
+        for (const auto [id, name] : _windowManager.GetPeasantNames())
+        {
+            winrt::hstring displayText = name;
+            if (name.empty())
+            {
+                displayText = fmt::format(L"{} {} - <{}>", locWindow, id, locUnnamed);
+            }
+
+            AppendMenu(hmenu, MF_STRING, (UINT_PTR)id, displayText.c_str());
+        }
+        return hmenu;
+    }
+    return nullptr;
+}
+
+// Method Description:
+// - This is the handler for when one of the menu items are selected within
+//   the tray icon's context menu.
+// Arguments:
+// - menu: The handle to the menu that holds the menu item that was selected.
+// - menuItemIndex: The index of the menu item within the given menu.
+// Return Value:
+// - <none>
+void AppHost::_TrayMenuItemSelected(const HMENU menu, const UINT menuItemIndex)
+{
+    // Let's find out which menu/submenu we're looking at.
+    MENUINFO mi{};
+    mi.cbSize = sizeof(MENUINFO);
+    mi.fMask = MIM_MENUDATA;
+    GetMenuInfo(menu, &mi);
+
+    if (mi.dwMenuData)
+    {
+        if ((TrayMenuItemAction)mi.dwMenuData == TrayMenuItemAction::SummonWindow)
+        {
+            Remoting::SummonWindowSelectionArgs args{};
+            args.WindowID(GetMenuItemID(menu, menuItemIndex));
+            args.SummonBehavior().ToggleVisibility(false);
+            _windowManager.SummonWindow(args);
+        }
+    }
+
+    auto action = (TrayMenuItemAction)GetMenuItemID(menu, menuItemIndex);
+    switch (action)

```

Comment:
surprised you don't need a case for `SummonWindow` or the default case

Reply:
So I'm looking for `SummonWindow` right above the switch statement because I'm expecting `SummonWindow` to be set as the data of the _owning menu_ for the window _menu items_. Basically I'm trying to say that everything in this menu should be treated as a `SummonWindow` action, and this way I can place the ID of the window in each menu item to know which one to summon.

The switch statement on the other hand is used to handle the action of a _menu item_, aka. `Focus Terminal` or any future menu items that do a specific action.

I don't believe there's a default case though - I wouldn't want to do perform an action if they click on something that doesn't have an action associated with it.